File: cli/internal/graphvisualizer/graphvisualizer.go
package graphvisualizer

import (
	"fmt"
	"io"
	"math/rand"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"

	"github.com/fatih/color"
	"github.com/mitchellh/cli"
	"github.com/pyr-sh/dag"
	"github.com/vercel/turbo/cli/internal/turbopath"
	"github.com/vercel/turbo/cli/internal/ui"
	"github.com/vercel/turbo/cli/internal/util"
	"github.com/vercel/turbo/cli/internal/util/browser"
)

// GraphVisualizer requirements
type GraphVisualizer struct {
	repoRoot  turbopath.AbsoluteSystemPath
	ui        cli.Ui
	TaskGraph *dag.AcyclicGraph
}

// hasGraphViz checks for the presence of https://graphviz.org/
func hasGraphViz() bool {
	err := exec.Command("dot", "-V").Run()
	return err == nil
}

func getRandChar() string {
	i := rand.Intn(25) + 65
	return string(rune(i))
}

func getRandID() string {
	return getRandChar() + getRandChar() + getRandChar() + getRandChar()
}

// New creates an instance of ColorCache with helpers for adding colors to task outputs
func New(repoRoot turbopath.AbsoluteSystemPath, ui cli.Ui, TaskGraph *dag.AcyclicGraph) *GraphVisualizer {
	return &GraphVisualizer{
		repoRoot:  repoRoot,
		ui:        ui,
		TaskGraph: TaskGraph,
	}
}

// Converts the TaskGraph dag into a string
func (g *GraphVisualizer) generateDotString() string {
	return string(g.TaskGraph.Dot(&dag.DotOpts{
		Verbose:    true,
		DrawCycles: true,
	}))
}

// Outputs a warning when a file was requested, but graphviz is not available
func (g *GraphVisualizer) graphVizWarnUI() {
	g.ui.Warn(color.New(color.FgYellow, color.Bold, color.ReverseVideo).Sprint(" WARNING ") + color.YellowString(" `turbo` uses Graphviz to generate an image of your\ngraph, but Graphviz isn't installed on this machine.\n\nYou can download Graphviz from https://graphviz.org/download.\n\nIn the meantime, you can use this string output with an\nonline Dot graph viewer."))
}

// RenderDotGraph renders a dot graph string for the current TaskGraph
func (g *GraphVisualizer) RenderDotGraph() {
	g.ui.Output("")
	g.ui.Output(g.generateDotString())
}

type nameCache map[string]string

func (nc nameCache) getName(in string) string {
	if existing, ok := nc[in]; ok {
		return existing
	}
	newName := getRandID()
	nc[in] = newName
	return newName
}

type sortableEdge dag.Edge
type sortableEdges []sortableEdge

// methods mostly copied from marshalEdges in the dag library
func (e sortableEdges) Less(i, j int) bool {
	iSrc := dag.VertexName(e[i].Source())
	jSrc := dag.VertexName(e[j].Source())
	if iSrc < jSrc {
		return true
	} else if iSrc > jSrc {
		return false
	}
	return dag.VertexName(e[i].Target()) < dag.VertexName(e[j].Target())
}
func (e sortableEdges) Len() int      { return len(e) }
func (e sortableEdges) Swap(i, j int) { e[i], e[j] = e[j], e[i] }

func (g *GraphVisualizer) generateMermaid(out io.StringWriter) error {
	if _, err := out.WriteString("graph TD\n"); err != nil {
		return err
	}
	cache := make(nameCache)
	// cast edges to our custom type so we can sort them
	// this allows us to generate the same graph every time
	var edges sortableEdges
	for _, edge := range g.TaskGraph.Edges() {
		edges = append(edges, sortableEdge(edge))
	}
	sort.Sort(edges)
	for _, edge := range edges {
		left := dag.VertexName(edge.Source())
		right := dag.VertexName(edge.Target())
		leftName := cache.getName(left)
		rightName := cache.getName(right)
		if _, err := out.WriteString(fmt.Sprintf("\t%v(\"%v\") --> %v(\"%v\")\n", leftName, left, rightName, right)); err != nil {
			return err
		}
	}
	return nil
}

// GenerateGraphFile saves a visualization of the TaskGraph to a file (or renders a DotGraph as a fallback))
func (g *GraphVisualizer) GenerateGraphFile(outputName string) error {
	outputFilename := g.repoRoot.UntypedJoin(outputName)
	ext := outputFilename.Ext()
	// use .jpg as default extension if none is provided
	if ext == "" {
		ext = ".jpg"
		outputFilename = g.repoRoot.UntypedJoin(outputName + ext)
	}
	if ext == ".mermaid" {
		f, err := outputFilename.Create()
		if err != nil {
			return fmt.Errorf("error creating file: %w", err)
		}
		defer util.CloseAndIgnoreError(f)
		if err := g.generateMermaid(f); err != nil {
			return err
		}
		g.ui.Output(fmt.Sprintf("✔ Generated task graph in %s", ui.Bold(outputFilename.ToString())))
		return nil
	}
	graphString := g.generateDotString()
	if ext == ".html" {
		f, err := outputFilename.Create()
		if err != nil {
			return fmt.Errorf("error creating file: %w", err)
		}
		defer f.Close() //nolint errcheck
		_, writeErr1 := f.WriteString(`<!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <title>Graph</title>
    </head>
    <body>
      <script src="https://cdn.jsdelivr.net/npm/viz.js@2.1.2-pre.1/viz.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/viz.js@2.1.2-pre.1/full.render.js"></script>
      <script>`)
		if writeErr1 != nil {
			return fmt.Errorf("error writing graph contents: %w", writeErr1)
		}

		_, writeErr2 := f.WriteString("const s = `" + graphString + "`.replace(/\\_\\_\\_ROOT\\_\\_\\_/g, \"Root\").replace(/\\[root\\]/g, \"\");new Viz().renderSVGElement(s).then(el => document.body.appendChild(el)).catch(e => console.error(e));")
		if writeErr2 != nil {
			return fmt.Errorf("error creating file: %w", writeErr2)
		}

		_, writeErr3 := f.WriteString(`
    </script>
  </body>
  </html>`)
		if writeErr3 != nil {
			return fmt.Errorf("error creating file: %w", writeErr3)
		}

		g.ui.Output("")
		g.ui.Output(fmt.Sprintf("✔ Generated task graph in %s", ui.Bold(outputFilename.ToString())))
		if ui.IsTTY {
			if err := browser.OpenBrowser(outputFilename.ToString()); err != nil {
				g.ui.Warn(color.New(color.FgYellow, color.Bold, color.ReverseVideo).Sprintf("failed to open browser. Please navigate to file://%v", filepath.ToSlash(outputFilename.ToString())))
			}
		}
		return nil
	}
	hasDot := hasGraphViz()
	if hasDot {
		dotArgs := []string{"-T" + ext[1:], "-o", outputFilename.ToString()}
		cmd := exec.Command("dot", dotArgs...)
		cmd.Stdin = strings.NewReader(graphString)
		if err := cmd.Run(); err != nil {
			return fmt.Errorf("could not generate task graphfile %v:  %w", outputFilename, err)
		}
		g.ui.Output("")
		g.ui.Output(fmt.Sprintf("✔ Generated task graph in %s", ui.Bold(outputFilename.ToString())))

	} else {
		g.ui.Output("")
		// User requested a file, but we're falling back to console here so warn about installing graphViz correctly
		g.graphVizWarnUI()
		g.RenderDotGraph()
	}
	return nil
}

File: crates/turborepo-lib/src/run/graph_visualizer.rs
use std::{
    fs::OpenOptions,
    io::{self, Write},
    process::{Command, Stdio},
};

use thiserror::Error;
use turbopath::{AbsoluteSystemPath, AbsoluteSystemPathBuf};
use turborepo_ui::{cprintln, cwrite, cwriteln, ColorConfig, BOLD, BOLD_YELLOW_REVERSE, YELLOW};
use which::which;

use crate::{engine::Engine, opts::GraphOpts, spawn_child};

#[derive(Debug, Error)]
pub enum Error {
    #[error("failed to produce graph output: {0}")]
    GraphOutput(#[source] std::io::Error),
    #[error("invalid graph filename {raw_filename}: {reason}")]
    InvalidFilename {
        raw_filename: String,
        reason: String,
    },
    #[error("failed to spawn graphviz (dot): {0}")]
    Graphviz(io::Error),
}

pub(crate) fn write_graph(
    ui: ColorConfig,
    graph_opts: &GraphOpts,
    engine: &Engine,
    single_package: bool,
    cwd: &AbsoluteSystemPath,
) -> Result<(), Error> {
    match graph_opts {
        GraphOpts::Stdout => render_dot_graph(std::io::stdout(), engine, single_package)?,
        GraphOpts::File(raw_filename) => {
            let (filename, extension) = filename_and_extension(cwd, raw_filename)?;
            if extension == "mermaid" {
                render_mermaid_graph(&filename, engine, single_package)?;
            } else if extension == "html" {
                render_html(&filename, engine, single_package)?;
            } else if let Ok(dot_path) = which("dot") {
                let mut cmd = Command::new(dot_path);
                cmd.stdin(Stdio::piped())
                    .args(["-T", extension.as_str(), "-o", filename.as_str()])
                    .current_dir(cwd);
                let child = spawn_child(cmd).map_err(Error::Graphviz)?;
                let stdin = child.take_stdin().expect("graphviz should have a stdin");
                render_dot_graph(stdin, engine, single_package)?;
                child.wait().map_err(Error::Graphviz)?;
            } else {
                write_graphviz_warning(ui).map_err(Error::GraphOutput)?;
                render_dot_graph(std::io::stdout(), engine, single_package)?;
            }
            print!("\n✓ Generated task graph in ");
            cprintln!(ui, BOLD, "{filename}");
        }
    }
    Ok(())
}

fn write_graphviz_warning(color_config: ColorConfig) -> Result<(), io::Error> {
    let stderr = io::stderr();
    cwrite!(&stderr, color_config, BOLD_YELLOW_REVERSE, " WARNING ")?;
    cwriteln!(&stderr, color_config, YELLOW, " `turbo` uses Graphviz to generate an image of your\ngraph, but Graphviz isn't installed on this machine.\n\nYou can download Graphviz from https://graphviz.org/download.\n\nIn the meantime, you can use this string output with an\nonline Dot graph viewer.")?;
    Ok(())
}

fn render_mermaid_graph(
    filename: &AbsoluteSystemPath,
    engine: &Engine,
    single_package: bool,
) -> Result<(), Error> {
    let mut opts = OpenOptions::new();
    opts.truncate(true).create(true).write(true);
    let file = filename
        .open_with_options(opts)
        .map_err(Error::GraphOutput)?;
    engine
        .mermaid_graph(file, single_package)
        .map_err(Error::GraphOutput)
}

fn render_dot_graph<W: io::Write>(
    writer: W,
    engine: &Engine,
    single_package: bool,
) -> Result<(), Error> {
    engine
        .dot_graph(writer, single_package)
        .map_err(Error::GraphOutput)
}

const HTML_PREFIX: &str = r#"
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <title>Graph</title>
</head>
<body>
  <script src="https://cdn.jsdelivr.net/npm/viz.js@2.1.2-pre.1/viz.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/viz.js@2.1.2-pre.1/full.render.js"></script>
  <script>
"#;
const HTML_SUFFIX: &str = r#"
  </script>
</body>
</html>
"#;

fn render_html(
    filename: &AbsoluteSystemPath,
    engine: &Engine,
    single_package: bool,
) -> Result<(), Error> {
    let mut opts = OpenOptions::new();
    opts.truncate(true).create(true).write(true);
    let mut file = filename
        .open_with_options(opts)
        .map_err(Error::GraphOutput)?;
    let mut graph_buffer = Vec::new();
    render_dot_graph(&mut graph_buffer, engine, single_package)?;
    let graph_string = String::from_utf8(graph_buffer).expect("graph rendering should be UTF-8");

    file.write_all(HTML_PREFIX.as_bytes())
        .map_err(Error::GraphOutput)?;
    write!(
        &mut file,
        "const s = `{graph_string}`.replace(/\\_\\_\\_ROOT\\_\\_\\_/g, \
         \"Root\").replace(/\\[root\\]/g, \"\");new Viz().renderSVGElement(s).then(el => \
         document.body.appendChild(el)).catch(e => console.error(e));"
    )
    .map_err(Error::GraphOutput)?;
    file.write_all(HTML_SUFFIX.as_bytes())
        .map_err(Error::GraphOutput)?;
    Ok(())
}

fn filename_and_extension(
    cwd: &AbsoluteSystemPath,
    raw_filename: &str,
) -> Result<(AbsoluteSystemPathBuf, String), Error> {
    let graph_file = AbsoluteSystemPathBuf::from_unknown(cwd, raw_filename);
    if let Some(extension) = graph_file.extension() {
        let extension = extension.to_string();
        Ok((graph_file, extension))
    } else {
        let extension = "jpg".to_string();
        let filename = graph_file
            .file_name()
            .ok_or_else(|| Error::InvalidFilename {
                raw_filename: raw_filename.to_string(),
                reason: "Cannot get filename from path".to_string(),
            })?;
        let jpg_filename = format!("{filename}.{extension}");
        let jpg_graph_file = graph_file
            .parent()
            .ok_or_else(|| Error::InvalidFilename {
                raw_filename: raw_filename.to_string(),
                reason: "Cannot get parent of output file".to_string(),
            })?
            .join_component(&jpg_filename);
        Ok((jpg_graph_file, extension))
    }
}
